File: cpp/day_13_Largest Divisible Subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int i, j, n, ans = 0;
        n = nums.size();
        if(n < 1){
            return {};
        }
        vector<int> dp(n, 1), path(n, -1), res;
        sort(nums.begin(), nums.end());
        for(i = 1; i < n; i++){
            for(j = 0; j < i; j++){
                if(nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i]){
                    dp[i] = dp[j] + 1;
                    path[i] = j;
                }
            }
            if(dp[ans] < dp[i]){
                ans = i;
            }
        }
        j = ans;
        while(j >= 0){
            res.push_back(nums[j]);
            j = path[j];
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: python3/june/day_13_Largest Divisible Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        if n <= 1:
            return nums
        nums.sort()
        dp = [1] * n
        path = [-1] * n
        ans = 0
        for i in range(1, n):
            for j in range(0, i):
                if nums[i] % nums[j] == 0 and dp[i] < dp[j] + 1:
                    dp[i] = dp[j] + 1
                    path[i] = j
            if dp[i] > dp[ans]:
                ans = i
        
        res = []
        while ans >= 0:
            res.append(nums[ans])
            ans = path[ans]
        res.reverse()
        return res
